File: 프로그래머스/lv2/지준혁/오픈채팅방.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>
#include <sstream>

using namespace std;

vector<string> solution(vector<string> record) {
    
    unordered_map<string, string> map;
    
    for (auto e : record) {
        istringstream iss(e);
        vector<string> split;
        string tmp;
        while (getline(iss, tmp, ' ')) 
            split.push_back(tmp);
        if (split[0] == "Enter" || split[0] == "Change") {
            map[split[1]] = split[2];
        }
    }
    vector<string> ans;
    for (auto& e: record) {
        string ret = "";
        istringstream iss(e);
        vector<string> split;
        string tmp;
        while (getline(iss, tmp, ' ')) 
            split.push_back(tmp);
        
        if (split[0] == "Enter") {
            ret += map[split[1]];
            ret += "님이 들어왔습니다.";
            ans.push_back(ret);
        }
        else if (split[0] == "Leave") {
            ret += map[split[1]];
            ret += "님이 나갔습니다.";
            ans.push_back(ret);
        }
    }

    return ans;
}

File: 프로그래머스/lv2/이재표/오픈채팅방.java
import java.util.*;
class Solution {
    public String[] solution(String[] records) {
        Map<String,String>map=new HashMap<>();
        int cnt=0;
        for(int i=0;i<records.length;i++){
            String[] record=records[i].split(" ");
            if(record[0].equals("Change")){
                map.put(record[1],record[2]);
                continue;
            }
            if(record[0].equals("Enter")){
                map.put(record[1],record[2]);
            }
            cnt++;
        }
        String[] answer = new String[cnt];
        int idx=0;
        for(int i=0;i<records.length;i++){
            String[] record=records[i].split(" ");
            if(record[0].equals("Enter")){
                answer[idx++]=map.get(record[1])+"님이 들어왔습니다.";
            }else if(record[0].equals("Leave")){
                answer[idx++]=map.get(record[1])+"님이 나갔습니다.";
            }
        }
        return answer;
    }
}
